Client/Frontend/Browser/Authenticator.swift

```diff
@@ -48,6 +48,48 @@ class Authenticator {
         return self.promptForUsernamePassword(viewController, credentials: nil, protectionSpace: challenge.protectionSpace, loginsHelper: nil)
     }
 
+    static func findMatchingCredentialsForChallenge(challenge: NSURLAuthenticationChallenge, fromLoginsProvider loginsProvider: BrowserLogins) -> Deferred<NSURLCredential?> {
+        let deferred = Deferred<NSURLCredential?>()
+        loginsProvider.getLoginsForProtectionSpace(challenge.protectionSpace).upon { res in
+            var credentials: NSURLCredential? = nil
+            if let logins = res.successValue?.asArray() {
+
+                // It is possible that we might have duplicate entries since we match against host and scheme://host.
+                // This is a side effect of https://bugzilla.mozilla.org/show_bug.cgi?id=1238103.
+                if logins.count > 1 {
+                    credentials = (logins.find { login in
+                        (login.protectionSpace.`protocol` == challenge.protectionSpace.`protocol`) && !(login.hasMalformedHostname ?? false)
+                    })?.credentials
+
+                    let malformedGUIDs: [GUID] = logins.flatMap { login in
+                        if login.hasMalformedHostname {
+                            return login.guid
+                        }
+                        return nil
+                    }
+                    loginsProvider.removeLoginsWithGUIDs(malformedGUIDs).upon { log.debug("Removed malformed logins. Success :\($0.isSuccess)") }
+                }
+
+                // Found a single entry but the schemes don't match. This is a result of a schemeless entry that we
+                // saved in a previous iteration of the app so we need to migrate it. We only care about the
+                // the username/password so we can rewrite the scheme to be correct.
+                else if logins.count == 1 && logins[0].protectionSpace.`protocol` != challenge.protectionSpace.`protocol` {
+                    let login = logins[0]
+                    credentials = login.credentials
+                    let new = Login(credential: login.credentials, protectionSpace: challenge.protectionSpace)
+                    loginsProvider.updateLoginByGUID(login.guid, new: new, significant: true).value
+                }
+
+                // Found a single entry that matches the scheme and host - good to go.
+                else {
+                    credentials = logins[0].credentials
+                }
+            }
+            deferred.fill(credentials)
```

Comment:
You're taking a little bit of your own path here — you're conflating a database error (`upon` is given a `Maybe` with no `successValue`) with not finding the login.